cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,197 @@
+   constexpr size_type w_leader_lane{0};
+   const int w_lane_id = threadIdx.x % warp_size;
+
+   size_type w_valid_count{0};
+  
+   // current warp.
+   size_type w_cur = w_begin + w_id; 
+   auto w_active_mask = __ballot_sync(0xFFFF'FFFF, w_cur <= w_end);
+   while(w_cur <= w_end){
+      // absolute element index
+      size_type index = (w_cur * warp_size) + w_lane_id;
+      bool in_range = (index >= begin && index < end);
+
+      bool valid = true;
+      if(has_validity){
+         valid = in_range && filter(index) ? lhs.is_valid(index) : rhs.is_valid(index);
+      }
+
+      // do the copy if-else, but only if this element is valid in the column to be copied 
+      if(in_range && valid){ 
+         out.element<T>(index) = filter(index) ? lhs.element<T>(index) : rhs.element<T>(index);
+      }
+      
+      // update validity
+      if(has_validity){         
+         // the final validity mask for this warp
+         int w_mask = __ballot_sync(w_active_mask, valid);
+         // only one guy in the warp needs to update the mask and count
+         if(w_lane_id == 0){
+            out.set_mask_word(w_cur, w_mask);            
+            w_valid_count += __popc(w_mask);
+         }
+      }      
+
+      // next grid
+      w_cur += blockDim.x * gridDim.x;
+      if(has_validity){
+         w_active_mask = __ballot_sync(w_active_mask, w_cur < w_end);
+      }
+   }
+
+   if(has_validity){
+      // sum all null counts across all warps
+      size_type b_valid_count = single_lane_block_sum_reduce<block_size, w_leader_lane>(w_valid_count);
+      // b_null_count will only be valid on thread 0
+      if(threadIdx.x == 0){
+         // using an atomic here because there are multiple blocks doing this work
+         atomicAdd(valid_count, b_valid_count);
+      }      
+   }
+}
+
+
+
+/* --------------------------------------------------------------------------*/
+/**
+* @brief Functor called by the `type_dispatcher` in order to perform a copy if/else
+*        using a filter function to select from lhs/rhs columns.
+*/
+/* ----------------------------------------------------------------------------*/
+struct copy_if_else_functor {
+   template <typename T, typename Filter>
+   void operator()(  column_view const& lhs,
+                     column_view const& rhs,
+                     Filter filter,
+                     mutable_column_view& out,
+                     rmm::mr::device_memory_resource *mr,                  
+                     cudaStream_t stream)
+   {
+      cudf::size_type num_els = warp_size * cudf::util::div_rounding_up_safe(lhs.size(), warp_size);
+      constexpr int block_size = 256;
+      cudf::experimental::detail::grid_1d grid{num_els, block_size, 1};
+
+      // device views
+      auto lhs_dv = column_device_view::create(lhs);
+      auto rhs_dv = column_device_view::create(rhs);
+      auto out_dv = mutable_column_device_view::create(out);
+      
+      if(lhs.nullable()){                           
```

Comment:
Ah, this is probably why you require both sides to have the same nullability. 
```suggestion
      if(lhs.has_nulls() or rhs.has_nulls()){                           
```
How about this instead.

Reply:
This is also incorrect. This only checks for the nullability of the `lhs`. You only need to check if it has nulls, and you need to check the `rhs` as well. Otherwise the bitmask updates will be incorrect.